lib/mdx.ml

```diff
@@ -73,31 +78,103 @@ let parse l =
       | `Block b   -> Block b
     ) l
 
+let parse_mli file_contents =
+  (* Find the locations of the code blocks within [file_contents], then slice it up into
+     [Text] and [Block] parts by using the starts and ends of those blocks as
+     boundaries. *)
+  let code_blocks = Mli_parser.docstring_code_blocks file_contents in
+  let cursor = ref { Odoc_model.Location_.line = 1; column = 0 } in
+  let tokens =
+    List.map
+      (fun (code_block : Mli_parser.Code_block.t) ->
+         let pre_text =
+           Text
+             (Mli_parser.slice
+                file_contents
+                ~start:!cursor
+                ~end_:code_block.location.start)
+         in
+         let block =
+           Block
+             { Block.line = code_block.location.start.line
+             ; file = ""
+             ; section = None
+             ; labels = []
+             ; header = Some "ocaml"
+             ; contents = Compat.String.split_on_char '\n' code_block.contents
+             ; value = Raw
+             }
+         in
+         cursor := code_block.location.end_;
+         [ pre_text; Text "{["; block; Text "]}" ])
+      code_blocks
+    |> List.concat
+  in
+  let eof =
+    let lines = Compat.String.split_on_char '\n' file_contents in
+    { Odoc_model.Location_.line = List.length lines
+    ; column = String.length (List.rev lines |> List.hd)
+    }
+  in
+  let eof_is_beyond_location (loc : Odoc_model.Location_.point) =
+    eof.line > loc.line || (eof.line = loc.line && eof.column > loc.column)
+  in
+  if eof_is_beyond_location !cursor
+  then (
+    let remainder = Mli_parser.slice file_contents ~start:!cursor ~end_:eof in
+    if not (Compat.String.equal remainder "") then tokens @ [ Text remainder ] else tokens)
+  else tokens
+
 type syntax = Syntax.t =
   | Normal
   | Cram
+  | Mli
+
+let parse_lexbuf file_contents syntax l =
+  match syntax with
+  | Syntax.Mli -> parse_mli file_contents
+  | Syntax.Normal | Syntax.Cram -> parse (Lexer_mdx.token syntax l)
 
-let parse_lexbuf syntax l = parse (Lexer.token syntax l)
-let parse_file syntax f = parse (Lexer.token syntax (snd (Misc.init f)))
-let of_string syntax s = parse_lexbuf syntax (Lexing.from_string s)
+let parse_file syntax f = parse (Lexer_mdx.token syntax (snd (Misc.init f)))
 
-let eval = function
+let of_string syntax s =
+  match syntax with
+  | Syntax.Mli -> parse_mli s
+  | Syntax.Normal | Syntax.Cram -> parse_lexbuf s syntax (Lexing.from_string s)
+
+let eval syntax = function
   | Section _ | Text _ as x -> x
   | Block t as x ->
-    let t' = Block.eval t in
+    let t' = Block.eval syntax t in
     if t == t' then x else Block t'
 
 type expect_result =
   | Identical
   | Differs
 
+let remove_whitespace str =
+  let open Astring in
+  List.fold_right (fun s acc ->
```

Comment:
This can be probably be simplified using `Astring`'s `String.filter Char.is_white`.

Could you also move it to `Util` please? That applies to other such functions you might have defined. The point is to avoid accidentally duplicate their code elsewhere if we ever need them.

Reply:
Sounds good, done! (Will push shortly.)